README.md
# mergebounce

`mergebounce` is a fork of lodash's [debounce](https://lodash.com/docs/4.17.15#debounce) function with the added feature that it'll use the lodash [merge](https://lodash.com/docs/4.17.15#merge)
function to merge all passed in parameters before eventually invoking the debounced inner function.

By doing so, it's possible to combine multiple expensive calls into a single one
without losing state changes that would have been made by individual calls.

## Install

```
npm i mergebounce
```

## usecases

### Batching expensive writes

Imagine you have a frontend app that stores data in IndexedDB via [localForage](https://localforage.github.io/localForage).
Writes to IndexedDB are slow, which can slow down your whole app.

You can increase performance by batching writes together.

One way to do this, is to install [localForage-setItems](https://github.com/localForage/localForage-setItems)
and then to use `mergebounce` to combine multple calls to `setItems` into a
single one.

```javascript
const debouncedSetItems = mergebounce(
    items => store.setItems(items),
    50,
    {'promise': true}
);

function save (json) {
    const items = {}
    items[json.id)] = json;
    return debouncedSetItems(items);
}
```

Now `save` can be called many times, but the actual writes (calls to
`setItems`) will be far fewer.


### Reducing requests

Let's suppose you have a chat app and as chat messages appear you want to fetch
user data for new message authors.

When you initially load the chat, there might be a 100 messages with 50
authors.

Instead of making a request for every incoming message that has an
as yet unknown author, you can mergebounce the function and combine multiple
calls into a single request.

For example:

```javascript
async function _fetchUserData(nicknames) {
  const response = await fetch('/user/data', {
    body: JSON.stringify({nicknames}),
    headers: { 'Content-Type': 'application/json' },
  });

  data.forEach(userdata => getUser().update(userdata));
}

const fetchUserData = mergebounce(_fetchUserData, 250, {'concatArrays': true});

// The following calls with result in one request with all the nicnames
// concatenated into one array
fetchUserData(['coolguy69', 'ilikecats', 'dielan']);
fetchUserData(['coolboymew']);
fetchUserData(['donkeykong']);
```

## options

The default `debounce` options are allowed, as well as the following option:

* `concatArrays`:
    By default arrays will be treated as objects when being merged. When
    merging two arrays, the values in the 2nd arrray will replace the
    corresponding values (i.e. those with the same indexes) in the first array.
    When `concatArrays` is set to `true`, arrays will be concatenated instead.
* `dedupeArrays`:
    This option is similar to `concatArrays`, except that the concatenated
    array will also be deduplicated. Thus any entries that are concatenated to the
    existing array, which are already contained in the existing array, will
    first be removed.
* `promise`:
    By default, when calling a mergebounced function that doesn't execute
    immediately, you'll receive the result from its previous execution, or
    `undefined` if it has never executed before. By setting the `promise`
    option to `true`, a promise will be returned instead of the previous
    execution result when the function is debounced. The promise will resolve
    with the result of the next execution, as soon as it happens.


CHANGES.md
# Changelog

## 0.1.1 (2022-04-08)

- Bugfix. Return resolved promise when invoking the function.

## 0.1.0 (2021-07-08)

- Add new option `dedupeArrays`, similar to `concatArrays` but duplicate
  entries are removed.

## 0.0.3 (2021-06-30)

- Use babel to transpile the source files. Solves issues with optional chaining
  not yet supported everywhere.

## 0.0.2 (2021-04-28)

- Add new options `concatArrays` and `promise`

## 0.0.1 (2021-03)

- Initial release


tests.js
/*global describe, it, expect, mergebounce */

describe("mergebounce", function () {

  it("merges passed in parameters before invoking the function", function (done) {
    let invokedData;
    const mergebounced = mergebounce(data => (invokedData = data), 10);
    mergebounced({'a': [{ 'b': 2 }, { 'd': 4 }]});
    mergebounced({'a': [{ 'c': 3 }, { 'e': 5 }]});

    setTimeout(() => {
      expect(invokedData).toEqual({ 'a': [{ 'b': 2, 'c': 3 }, { 'd': 4, 'e': 5 }] })
      done();
    }, 15);
  });

  it("replaces arrays that don't contain objects", function (done) {
    let invokedData;
    const mergebounced = mergebounce(data => (invokedData = data), 10);
    mergebounced([{a: 2}, 4, 5]);
    mergebounced([{a: 1}, 2, 3, 4]);

    setTimeout(() => {
      expect(invokedData).toEqual([{a: 1}, 2, 3, 4]);
      done();
    }, 15);
  });

  it("concatenates arrays if concatArrays option is set", function (done) {
    let invokedData;
    const mergebounced = mergebounce(data => (invokedData = data), 10, {'concatArrays': true});
    mergebounced([{a: 2}, 4, 5]);
    mergebounced([{a: 1}, 2, 3, 4]);

    setTimeout(() => {
      expect(invokedData).toEqual([{a: 2}, 4, 5, {a: 1}, 2, 3, 4]);
      done();
    }, 15);
  });

  it('should dedupe concatenated arrays if "dedupeArrays" is set to true', function (done) {
    let callCount = 0;
    const values = [];

    let mergebounced = mergebounce((value) => {
      ++callCount;
      values.push(value);
    }, 32, {'dedupeArrays': true});

    mergebounced(['a']);
    mergebounced(['a']);
    mergebounced(['b']);
    expect(values.length).toBe(0);
    expect(callCount).toBe(0);

    setTimeout(() => {
      expect(callCount).toBe(1);
      expect(values).toEqual([['a', 'b']]);
      done();
    }, 128);
  });


  it("returns a promise if the promise option is set to true", function (done) {
    const mergebounced = mergebounce(() => {}, 10, {'promise': true});
    const result = mergebounced([{a: 2}, 4, 5]);
    expect(result instanceof Promise).toBe(true);
    expect(result.isResolved).toBe(false);
    setTimeout(() => {
      expect(result.isResolved).toBe(true);
      done();
    }, 15);
  });

  it("merges nested arrays uniquely", function (done) {
    let invokedData;
    const mergebounced = mergebounce(data => (invokedData = data), 10);
    mergebounced({'a': [{ 'b': 2 }, { 'd': [{a: 4}, {b: 6}] }]});
    mergebounced({'a': [{ 'c': 3 }, { 'd': [{a: 5}] }]});

    setTimeout(() => {
      expect(invokedData).toEqual({ 'a': [{ 'b': 2, 'c': 3 }, { 'd': [{a: 5}, {b: 6}] }] })
      done();
    }, 15);
  });

  it("gracefully handles no arguments being passed in", function (done) {
    let invokedData;
    const mergebounced = mergebounce(data => (invokedData = data), 10);
    mergebounced({'a': [{ 'b': 2 }, { 'd': [4] }]});
    mergebounced();

    setTimeout(() => {
      expect(invokedData).toEqual({'a': [{ 'b': 2 }, { 'd': [4] }]})
      done();
    }, 15);
  });

  it("handles multple arguments being passed in", function (done) {
    let args;
    const mergebounced = mergebounce(function () {
      args = Array.from(arguments);
    }, 10);
    mergebounced({ 'b': 2 }, { 'd': [4] });
    mergebounced({ 'c': 3 }, { 'd': [4, 5]});

    setTimeout(() => {
      expect(args).toEqual([{ 'b': 2, 'c': 3 }, { 'd': [4, 5] }])
      done();
    }, 15);
  });

  it("handles multple arguments of variable length", function (done) {
    let args;
    const mergebounced = mergebounce(function () {
      args = Array.from(arguments);
    }, 10);
    mergebounced({ 'b': 2 }, { 'd': [4] });
    mergebounced({ 'c': 3 });

    setTimeout(() => {
      expect(args).toEqual([{ 'b': 2, 'c': 3 }, { 'd': [4] }])
      done();
    }, 15);
  });

  it("keeps track of multple mergebounce functions", function (done) {
    let args1, args2;
    const mergebounced = mergebounce(function () {
      args1 = Array.from(arguments);
    }, 10);

    const mergebounced2 = mergebounce(function () {
      args2 = Array.from(arguments);
    }, 10);

    mergebounced({ 'b': 2 }, { 'd': [4] });
    mergebounced2({'a': [{ 'b': 2 }, { 'd': [4] }]});

    mergebounced({ 'c': 3 });
    mergebounced2({'a': [{ 'c': 3 }, { 'd': [4, 5] }]});

    setTimeout(() => {
      expect(args1).toEqual([{ 'b': 2, 'c': 3 }, { 'd': [4] }])
      expect(args2).toEqual([{ 'a': [{ 'b': 2, 'c': 3 }, { 'd': [4, 5] }] }])
      done();
    }, 15);
  });

  it('should debounce a function', function (done) {
    let callCount = 0;

    let mergebounced = mergebounce(function (value) {
      ++callCount;
      return value;
    }, 32);

    let results = [mergebounced('a'), mergebounced('b'), mergebounced('c')];
    expect(results).toEqual([undefined, undefined, undefined]);
    expect(callCount).toBe(0);

    setTimeout(function() {
      expect(callCount).toBe(1);
      const results = [mergebounced(['d']), mergebounced(['e']), mergebounced(['f'])];
      expect(results).toEqual(['c', 'c', 'c']);
      expect(callCount).toBe(1);
    }, 128);

    setTimeout(function() {
      expect(callCount).toBe(2);
      done();
    }, 256);
  });

  it("should not flush if it wasn't called at least once", function (done) {
    let callCount = 0;
    const mergebounced = mergebounce(() => ++callCount, 15);
    mergebounced.flush();
    expect(callCount).toBe(0);
    done();
  });

  it('should resolve when flushing', function(done) {
    let callCount = 0;
    const mergebounced = mergebounce(() => ++callCount, 0, {promise: true});

    mergebounced();
    mergebounced();
    expect(callCount).toBe(0);

    mergebounced.flush().then(() => {
      expect(callCount).toBe(1);
      done();
    });
  });

  it('should not immediately call `func` when `wait` is `0`', function(done) {
    let callCount = 0;
    const mergebounced = mergebounce(function() { ++callCount; }, 0);

    mergebounced();
    mergebounced();
    expect(callCount).toBe(0);

    setTimeout(function() {
      expect(callCount).toBe(1);
      done();
    }, 5);
  });

  it('should apply default options', function(done) {
    let callCount = 0;
    const mergebounced = mergebounce(() => callCount++, 32, {});

    mergebounced();
    expect(callCount).toEqual(0);

    setTimeout(function() {
      expect(callCount).toEqual(1);
      done();
    }, 64);
  });

  it('should support a `maxWait` option', function(done) {
    let callCount = 0;

    const mergebounced = mergebounce(function(value) {
      ++callCount;
      return value;
    }, 32, { 'maxWait': 64 });

    mergebounced();
    mergebounced();
    expect(callCount).toEqual(0);

    setTimeout(function() {
      expect(callCount).toEqual(1);
      mergebounced();
      mergebounced();
      expect(callCount).toEqual(1);
    }, 128);

    setTimeout(function() {
      expect(callCount).toEqual(2);
      done();
    }, 256);
  });

  it('should support `maxWait` in a tight loop', function(done) {
    let limit = 320,
        withCount = 0,
        withoutCount = 0;

    const withMaxWait = mergebounce(function() {
      withCount++;
    }, 64, { 'maxWait': 128 });

    let withoutMaxWait = mergebounce(function() {
      withoutCount++;
    }, 96);

    let start = +new Date;
    while ((new Date - start) < limit) {
      withMaxWait();
      withoutMaxWait();
    }
    let actual = [Boolean(withoutCount), Boolean(withCount)];
    setTimeout(function() {
      expect(actual).toEqual([false, true]);
      done();
    }, 1);
  });

  it('should queue a trailing call for subsequent mergebounced calls after `maxWait`', function(done) {
    let callCount = 0;

    let mergebounced = mergebounce(() => ++callCount, 200, { 'maxWait': 200 });

    mergebounced();

    setTimeout(mergebounced, 190);
    setTimeout(mergebounced, 200);
    setTimeout(mergebounced, 210);

    setTimeout(function() {
      expect(callCount).toEqual(2);
      done();
    }, 500);
  });

  it('should cancel `maxDelayed` when `delayed` is invoked', function(done) {
    let callCount = 0;

    let mergebounced = mergebounce(function() {
      callCount++;
    }, 32, { 'maxWait': 64 });

    mergebounced();

    setTimeout(function() {
      mergebounced();
      expect(callCount).toEqual(1);
    }, 128);

    setTimeout(function() {
      expect(callCount).toEqual(2);
      done();
    }, 192);
  });
});


karma.conf.js
// Karma configuration
// Generated on Thu Mar 18 2021 11:07:09 GMT+0100 (Central European Standard Time)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],

    // list of files / patterns to load in the browser
    files: [
      './dist/mergebounce.js',
      './tests.js'
    ],

    // list of files / patterns to exclude
    exclude: [],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {},

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress', 'kjhtml'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}


stashbounce.js
import isObject from 'lodash-es/isObject.js';
import merge from 'lodash-es/merge.js';
import now from 'lodash-es/now.js';
import toNumber from 'lodash-es/toNumber.js';

/** Error message constants. */
const FUNC_ERROR_TEXT = 'Expected a function';

/* Built-in method references for those with the same name as other `lodash` methods. */
const nativeMax = Math.max;
const nativeMin = Math.min;

/**
 * Creates a debounced function that delays invoking `func` until after `wait`
 * milliseconds have elapsed since the last time the debounced function was
 * invoked. The debounced function comes with a `cancel` method to cancel
 * delayed `func` invocations and a `flush` method to immediately invoke them.
 *
 * This function differs from lodash's debounce by merging all passed objects
 * before passing them to the final invoked function.
 *
 * Because of this, invoking can only happen on the trailing edge, since
 * passed-in data would be discarded if invoking happened on the leading edge.
 *
 * If `wait` is `0`, `func` invocation is deferred until to the next tick,
 * similar to `setTimeout` with a timeout of `0`.
 *
 * @static
 * @category Function
 * @param {Function} func The function to stashbounce.
 * @param {number} [wait=0] The number of milliseconds to delay.
 * @param {Object} [options={}] The options object.
 * @param {number} [options.maxWait]
 *  The maximum time `func` is allowed to be delayed before it's invoked.
 * @returns {Function} Returns the new debounced function.
 * @example
 *
 * // Avoid costly calculations while the window size is in flux.
 * window.addEventListener('resize', stashbounce(calculateLayout, 150));
 *
 * // Invoke `sendMail` when clicked, debouncing subsequent calls.
 * element.addEventListner('click', stashbounce(sendMail, 300));
 *
 * // Ensure `batchLog` is invoked once after 1 second of debounced calls.
 * const stashbounced = stashbounce(batchLog, 250, { 'maxWait': 1000 });
 * const source = new EventSource('/stream');
 * jQuery(source).on('message', stashbounced);
 *
 * // Cancel the trailing debounced invocation.
 * window.addEventListener('popstate', stashbounced.cancel);
 */
function stashbounce(func, wait, options) {
  let lastArgs = [],
      lastThis,
      maxWait,
      result,
      timerId,
      lastCallTime,
      lastInvokeTime = 0,
      maxing = false;

  if (typeof func != 'function') {
    throw new TypeError(FUNC_ERROR_TEXT);
  }
  wait = toNumber(wait) || 0;
  if (isObject(options)) {
    maxing = 'maxWait' in options;
    maxWait = maxing ? nativeMax(toNumber(options.maxWait) || 0, wait) : maxWait;
  }

  function invokeFunc(time) {
    const args = lastArgs;
    const thisArg = lastThis;
    lastArgs = [];
    lastThis = undefined;
    lastInvokeTime = time;
    result = func.apply(thisArg, args);
    return result;
  }

  function leadingEdge(time) {
    // Reset any `maxWait` timer.
    lastInvokeTime = time;
    // Start the timer for the trailing edge.
    timerId = setTimeout(timerExpired, wait);
    return result;
  }

  function remainingWait(time) {
    const timeSinceLastCall = time - lastCallTime;
    const timeSinceLastInvoke = time - lastInvokeTime;
    const timeWaiting = wait - timeSinceLastCall;
    return maxing
      ? nativeMin(timeWaiting, maxWait - timeSinceLastInvoke)
      : timeWaiting;
  }

  function shouldInvoke(time) {
    const timeSinceLastCall = time - lastCallTime;
    const timeSinceLastInvoke = time - lastInvokeTime;

    // Either this is the first call, activity has stopped and we're at the
    // trailing edge, the system time has gone backwards and we're treating
    // it as the trailing edge, or we've hit the `maxWait` limit.
    return (lastCallTime === undefined || (timeSinceLastCall >= wait) ||
      (timeSinceLastCall < 0) || (maxing && timeSinceLastInvoke >= maxWait));
  }

  function timerExpired() {
    const time = now();
    if (shouldInvoke(time)) {
      return trailingEdge(time);
    }
    // Restart the timer.
    timerId = setTimeout(timerExpired, remainingWait(time));
  }

  function trailingEdge(time) {
    timerId = undefined;

    // Only invoke if we have `lastArgs` which means `func` has been
    // debounced at least once.
    if (lastArgs) {
      return invokeFunc(time);
    }
    lastArgs = [];
    lastThis = undefined;
    return result;
  }

  function cancel() {
    if (timerId !== undefined) {
      clearTimeout(timerId);
    }
    lastInvokeTime = 0;
    lastArgs = [];
    lastCallTime = lastThis = timerId = undefined;
  }

  function flush() {
    return timerId === undefined ? result : trailingEdge(now());
  }

  function mergeArguments(args) {
    return args.reduce((acc, newValue) => {
      if (!newValue) {
        return acc;
      }
      const idx = args.indexOf(newValue);
      const oldValue = lastArgs[idx];
      let mergedValue;
      if (Array.isArray(oldValue) && Array.isArray(newValue)) {
        mergedValue = [...oldValue, ...newValue];
      } else if (isObject(oldValue) && isObject(newValue)) {
        mergedValue = merge(oldValue, newValue);
      }
      [...acc.splice(0, idx), mergedValue, ...args.splice(idx+1)]
    }, []);
  }

  function debounced() {
    const time = now();
    const isInvoking = shouldInvoke(time);

    lastArgs = mergeArguments(Array.from(arguments));
    lastThis = this;
    lastCallTime = time;

    if (isInvoking) {
      if (timerId === undefined) {
        return leadingEdge(lastCallTime);
      }
      if (maxing) {
        // Handle invocations in a tight loop.
        clearTimeout(timerId);
        timerId = setTimeout(timerExpired, wait);
        return invokeFunc(lastCallTime);
      }
    }
    if (timerId === undefined) {
      timerId = setTimeout(timerExpired, wait);
    }
    return result;
  }
  debounced.cancel = cancel;
  debounced.flush = flush;
  return debounced;
}

export default stashbounce;


mergebounce.js
import isObject from 'lodash-es/isObject.js';
import merge from 'lodash-es/merge.js';
import mergeWith from 'lodash-es/mergeWith.js';
import now from 'lodash-es/now.js';
import toNumber from 'lodash-es/toNumber.js';
import { getOpenPromise } from '@converse/openpromise/openpromise.js';

/** Error message constants. */
const FUNC_ERROR_TEXT = 'Expected a function';

/* Built-in method references for those with the same name as other `lodash` methods. */
const nativeMax = Math.max;
const nativeMin = Math.min;

/**
 * Creates a debounced function that delays invoking `func` until after `wait`
 * milliseconds have elapsed since the last time the debounced function was
 * invoked. The debounced function comes with a `cancel` method to cancel
 * delayed `func` invocations and a `flush` method to immediately invoke them.
 *
 * This function differs from lodash's debounce by merging all passed objects
 * before passing them to the final invoked function.
 *
 * Because of this, invoking can only happen on the trailing edge, since
 * passed-in data would be discarded if invoking happened on the leading edge.
 *
 * If `wait` is `0`, `func` invocation is deferred until to the next tick,
 * similar to `setTimeout` with a timeout of `0`.
 *
 * @static
 * @category Function
 * @param {Function} func The function to mergebounce.
 * @param {number} [wait=0] The number of milliseconds to delay.
 * @param {Object} [options={}] The options object.
 * @param {number} [options.maxWait]
 *  The maximum time `func` is allowed to be delayed before it's invoked.
 * @param {boolean} [options.concatArrays=false]
 *  By default arrays will be treated as objects when being merged. When
 *  merging two arrays, the values in the 2nd arrray will replace the
 *  corresponding values (i.e. those with the same indexes) in the first array.
 *  When `concatArrays` is set to `true`, arrays will be concatenated instead.
 * @param {boolean} [options.dedupeArrays=false]
 *  This option is similar to `concatArrays`, except that the concatenated
 *  array will also be deduplicated. Thus any entries that are concatenated to the
 *  existing array, which are already contained in the existing array, will
 *  first be removed.
 * @param {boolean} [options.promise=false]
 *  By default, when calling a merge-debounced function that doesn't execute
 *  immediately, you'll receive the result from its previous execution, or
 *  `undefined` if it has never executed before. By setting the `promise`
 *  option to `true`, a promise will be returned instead of the previous
 *  execution result when the function is debounced. The promise will resolve
 *  with the result of the next execution, as soon as it happens.
 * @returns {Function} Returns the new debounced function.
 * @example
 *
 * // Avoid costly calculations while the window size is in flux.
 * window.addEventListener('resize', mergebounce(calculateLayout, 150));
 *
 * // Invoke `sendMail` when clicked, debouncing subsequent calls.
 * element.addEventListner('click', mergebounce(sendMail, 300));
 *
 * // Ensure `batchLog` is invoked once after 1 second of debounced calls.
 * const mergebounced = mergebounce(batchLog, 250, { 'maxWait': 1000 });
 * const source = new EventSource('/stream');
 * jQuery(source).on('message', mergebounced);
 *
 * // Cancel the trailing debounced invocation.
 * window.addEventListener('popstate', mergebounced.cancel);
 */
function mergebounce(func, wait, options={}) {
  let lastArgs,
      lastThis,
      maxWait,
      result,
      timerId,
      lastCallTime,
      lastInvokeTime = 0,
      maxing = false;

  let promise = options.promise ? getOpenPromise() : null;

  if (typeof func != 'function') {
    throw new TypeError(FUNC_ERROR_TEXT);
  }
  wait = toNumber(wait) || 0;
  if (isObject(options)) {
    maxing = 'maxWait' in options;
    maxWait = maxing ? nativeMax(toNumber(options.maxWait) || 0, wait) : maxWait;
  }

  function invokeFunc(time) {
    const args = lastArgs;
    const thisArg = lastThis;
    const existingPromise = promise;
    lastArgs = lastThis = undefined;
    lastInvokeTime = time;
    result = func.apply(thisArg, args);
    if (options.promise) {
      existingPromise.resolve(result);
      promise = getOpenPromise();
    }
    return options.promise ? existingPromise : result;
  }

  function leadingEdge(time) {
    // Reset any `maxWait` timer.
    lastInvokeTime = time;
    // Start the timer for the trailing edge.
    timerId = setTimeout(timerExpired, wait);
    return options.promise ? promise : result;
  }

  function remainingWait(time) {
    const timeSinceLastCall = time - lastCallTime;
    const timeSinceLastInvoke = time - lastInvokeTime;
    const timeWaiting = wait - timeSinceLastCall;
    return maxing
      ? nativeMin(timeWaiting, maxWait - timeSinceLastInvoke)
      : timeWaiting;
  }

  function shouldInvoke(time) {
    const timeSinceLastCall = time - lastCallTime;
    const timeSinceLastInvoke = time - lastInvokeTime;

    // Either this is the first call, activity has stopped and we're at the
    // trailing edge, the system time has gone backwards and we're treating
    // it as the trailing edge, or we've hit the `maxWait` limit.
    return (lastCallTime === undefined || (timeSinceLastCall >= wait) ||
      (timeSinceLastCall < 0) || (maxing && timeSinceLastInvoke >= maxWait));
  }

  function timerExpired() {
    const time = now();
    if (shouldInvoke(time)) {
      return trailingEdge(time);
    }
    // Restart the timer.
    timerId = setTimeout(timerExpired, remainingWait(time));
  }

  function trailingEdge(time) {
    timerId = undefined;

    // Only invoke if we have `lastArgs` which means `func` has been
    // debounced at least once.
    if (lastArgs) {
      return invokeFunc(time);
    }
    lastArgs = lastThis = undefined;
    return options.promise ? promise : result;
  }

  function cancel() {
    if (timerId !== undefined) {
      clearTimeout(timerId);
    }
    lastInvokeTime = 0;
    lastArgs = lastCallTime = lastThis = timerId = undefined;
  }

  function flush() {
    return timerId === undefined ? result : trailingEdge(now());
  }

  function concatArrays(objValue, srcValue) {
    if (Array.isArray(objValue) && Array.isArray(srcValue)) {
      if (options?.dedupeArrays) {
        return objValue.concat(srcValue.filter(i => objValue.indexOf(i) === -1));
      } else {
        return objValue.concat(srcValue);
      }
    }
  }

  function mergeArguments(args) {
    if (lastArgs?.length) {
      if (!args.length) {
        return lastArgs;
      }
      if (options?.concatArrays || options?.dedupeArrays) {
        return mergeWith(lastArgs, args, concatArrays);
      } else {
        return merge(lastArgs, args);
      }
    } else {
      return args || [];
    }
  }

  function debounced() {
    const time = now();
    const isInvoking = shouldInvoke(time);

    lastArgs = mergeArguments(Array.from(arguments));
    lastThis = this;
    lastCallTime = time;

    if (isInvoking) {
      if (timerId === undefined) {
        return leadingEdge(lastCallTime);
      }
      if (maxing) {
        // Handle invocations in a tight loop.
        clearTimeout(timerId);
        timerId = setTimeout(timerExpired, wait);
        return invokeFunc(lastCallTime);
      }
    }
    if (timerId === undefined) {
      timerId = setTimeout(timerExpired, wait);
    }
    return options.promise ? promise : result;
  }
  debounced.cancel = cancel;
  debounced.flush = flush;
  return debounced;
}

export default mergebounce;
